src/gt4py/next/iterator/transforms/collapse_tuple.py

```diff
@@ -68,34 +140,68 @@ def apply(
         node: ir.Node,
         *,
         ignore_tuple_size: bool = False,
-        # the following options are mostly for allowing separate testing of the modes
-        collapse_make_tuple_tuple_get: bool = True,
-        collapse_tuple_get_make_tuple: bool = True,
         use_global_type_inference: bool = False,
+        # manually passing flags is mostly for allowing separate testing of the modes
+        flags=None,
     ) -> ir.Node:
         """
         Simplifies `make_tuple`, `tuple_get` calls.
 
         If `ignore_tuple_size`, apply the transformation even if length of the inner tuple
         is greater than the length of the outer tuple.
         """
-        node_types = it_type_inference.infer_all(node) if use_global_type_inference else None
-        return cls(
-            ignore_tuple_size,
-            collapse_make_tuple_tuple_get,
-            collapse_tuple_get_make_tuple,
-            use_global_type_inference,
-            node_types,
+        flags = flags or cls.flags
+        if use_global_type_inference:
+            it_type_inference.infer_all(node, save_to_annex=True)
+
+        new_node = cls(
+            ignore_tuple_size=ignore_tuple_size,
+            use_global_type_inference=use_global_type_inference,
+            flags=flags,
         ).visit(node)
 
-    def visit_FunCall(self, node: ir.FunCall, **kwargs) -> ir.Node:
-        if (
-            self.collapse_make_tuple_tuple_get
-            and node.fun == ir.SymRef(id="make_tuple")
-            and all(
-                isinstance(arg, ir.FunCall) and arg.fun == ir.SymRef(id="tuple_get")
-                for arg in node.args
+        # inline to remove left-overs from LETIFY_MAKE_TUPLE_ELEMENTS. this is important
+        # as otherwise two equal expressions containing a tuple will not be equal anymore
+        # and the CSE pass can not remove them.
+        # TODO: test case for `scan(lambda carry: {1, 2})` (see solve_nonhydro_stencil_52_like_z_q_tup)
+        if flags & cls.Flag.REMOVE_LETIFIED_MAKE_TUPLE_ELEMENTS:
+            new_node = InlineLambdas.apply(
+                new_node, opcount_preserving=True, force_inline_lambda_args=False
             )
+
+        return new_node
+
+    def visit_FunCall(self, node: ir.FunCall) -> ir.Node:
+        node = self.generic_visit(node)
+        return self.fp_transform(node)
+
+    def fp_transform(self, node: ir.Node) -> ir.Node:
+        while True:
+            new_node = self.transform(node)
+            if new_node is None:
+                break
+            assert new_node != node
+            node = new_node
+        return node
+
+    def transform(self, node: ir.Node) -> Optional[ir.Node]:
+        if not isinstance(node, ir.FunCall):
+            return None
+
+        for transformation in self.Flag:
+            if transformation == self.Flag.REMOVE_LETIFIED_MAKE_TUPLE_ELEMENTS:
+                continue
```

Comment:
Can you get rid of this? If it's not part of the default set, does it have to be accessible in the same pass?